### PR TITLE
InitialWorkDirRequirement part 2: StringDirent entryname expressions

### PR DESCRIPTION
**Pull request by cjllanwarne**
Motivating case: [link](https://github.com/broadinstitute/cromwell/blob/cjl_initial_work_dir_requirement_2/centaur/src/main/resources/standardTestCases/InitialWorkDirRequirement/inline_file_custom_entryname.cwl)

Allows the `entryname` to be an expression in `InitialWorkDirRequirement` `StringDirent` entries.

~~EDIT: Will rebase and target `develop` after merging https://github.com/broadinstitute/cromwell/pull/3078~~

### CODE REVIEW

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 1]
```diff
@@ -242,7 +242,8 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
 
     def adHocFileLocalization(womFile: WomFile): String = womFile.value.substring(womFile.value.lastIndexOf("/") + 1, womFile.value.length)
 
-    val adHocFileCreations: ErrorOr[List[CommandSetupSideEffectFile]] = jobDescriptor.taskCall.callable.adHocFileCreation.toList.traverse { _.evaluateValue(Map.empty, backendEngineFunctions) } map { _ collect {
+    val adHocFileCreationInputs = jobDescriptor.evaluatedTaskInputs.map { case (k,v) => k.localName.value -> v }
+    val adHocFileCreations: ErrorOr[List[CommandSetupSideEffectFile]] = jobDescriptor.taskCall.callable.adHocFileCreation.toList.traverse { _.evaluateValue(adHocFileCreationInputs, backendEngineFunctions) } map { _ collect {
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: preexisting, but my head would probably hurt a lot less if this was broken up into multiple lines with a temp var

File: cwl/src/main/scala/cwl/CwlWomExpression.scala [hunk 2]
```diff
@@ -106,6 +106,18 @@ final case class InitialWorkDirFileGeneratorExpression(entry: IwdrListingArrayEn
...
+        case other => WomStringType.coerceRawValue(other).map(_.asInstanceOf[WomString].value).toErrorOr
+      }
+      import common.validation.ErrorOr.ShortCircuitingFlatMap
+      for {
+        entryNameValidated <- entryNameEvaluation
+        writtenFile <- Try(Await.result(ioFunctionSet.writeFile(entryNameValidated, content), Duration.Inf)).toErrorOr
+      } yield writtenFile
+
+
     case _ => ??? // TODO WOM and the rest....
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/CwlWomExpression.scala]: 👁

File: cwl/src/main/scala/cwl/InitialWorkDirRequirement.scala [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
 
 final case class InitialWorkDirRequirement(
                                             `class`: W.`"InitialWorkDirRequirement"`.T,
-                                            listing: Array[IwdrListingArrayEntry] :+: ExpressionOrString :+: CNil
+                                            listing: Array[IwdrListingArrayEntry] :+: StringOrExpression :+: CNil
                                           ) {
   val listings: Array[IwdrListingArrayEntry] = listing match {
     case IwdrListingArray(array) => array
-    case ExpressionOrString(eos) => Array(Coproduct[IwdrListingArrayEntry](eos))
+    case StringOrExpression(eos) => Array(Coproduct[IwdrListingArrayEntry](eos))
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/InitialWorkDirRequirement.scala]: my ocd tells me the variable should get renamed too

File: cwl/src/main/scala/cwl/TypeAliases.scala [hunk 4]
```diff
@@ -97,6 +97,12 @@ trait TypeAliases {
   type ScatterVariables = Option[String :+: Array[String] :+: CNil]
 }
 
+object CwlAny {
+  implicit class EnhancedCwlAny(val cwlAny: CwlAny) extends AnyVal {
+    def stringRepresentation: String = cwlAny.select[Json] flatMap { json => json.asString } getOrElse "CwlAny (not Json)"
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/TypeAliases.scala]: sorry if this is a dumb q but can this really not be Json?

File: cwl/src/main/scala/cwl/package.scala [hunk 5]
```diff
@@ -60,6 +61,8 @@ package object cwl extends TypeAliases {
   type Expression = ECMAScriptExpression :+: ECMAScriptFunction :+: CNil
 
   object StringOrExpression {
+    def unapply(listing: Array[IwdrListingArrayEntry] :+: StringOrExpression :+: CNil): Option[StringOrExpression] = listing.select[StringOrExpression]
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/package.scala]: that coproduct feels like it should be a named `type`, which hopefully would make this custom `unapply` less surprising

File: cwl/src/main/scala/cwl/CwlWomExpression.scala (hunk 2 above)
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/CwlWomExpression.scala]: Still `TODO`, or maybe it won't be needed. Keep your eyes open for parts 3, 4, 5, ...

File: cwl/src/main/scala/cwl/TypeAliases.scala (hunk 4 above)
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/TypeAliases.scala]: I'm following the `CwlAny` type def. I suspect not (hence why I don't really bother trying to make it a string)

File: cwl/src/main/scala/cwl/package.scala (hunk 5 above)
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/package.scala]: Good idea, it also simplifies the `listing` field of the `InitialWorkDirRequirement` case class

**Pull request merged**